homeassistant/components/update/__init__.py

```diff
@@ -0,0 +1,297 @@
+
+
+class UpdateDescriptionUpdateCallback(Protocol):
+    """Protocol type for UpdateDescription.update_callback."""
+
+    def __call__(
+        self,
+        hass: HomeAssistant,
+        description: UpdateDescription,
+        **kwargs: Any,
+    ) -> Coroutine[None, None, bool]:
+        """Perform an update."""
+
+
+@dataclasses.dataclass()
+class UpdateDescription:
+    """Describe an update update."""
+
+    identifier: str
+    name: str
+    current_version: str
+    available_version: str
+    update_callback: UpdateDescriptionUpdateCallback
+    changelog_content: str | None = None
+    changelog_url: str | None = None
+    icon_url: str | None = None
+    provides_backup: bool = False
+
+
+@dataclasses.dataclass()
+class UpdateRegistration:
+    """Helper class to track platform registration."""
+
+    hass: HomeAssistant
+    domain: str
+    updates_callback: Callable[
+        [HomeAssistant], Awaitable[list[UpdateDescription]]
+    ] | None = None
+
+    @callback
+    def async_register(
+        self,
+        updates_callback: Callable[[HomeAssistant], Awaitable[list[UpdateDescription]]],
+    ):
+        """Register the updates info callback."""
+        manager: UpdateManager = self.hass.data[DOMAIN]
+        self.updates_callback = updates_callback
+        manager.register(self)
+
+
+class UpdateManager:
+    """Update manager for the update integration."""
+
+    def __init__(self, hass: HomeAssistant) -> None:
+        """Initialize the update manager."""
+        self._hass = hass
+        self._store = storage.Store(
+            hass=hass,
+            version=STORAGE_VERSION,
+            key=DOMAIN,
+        )
+        self._skip: set[str] = set()
+        self._updating: set[str] = set()
+        self._registrations: dict[str, UpdateRegistration] = {}
+        self._updates: dict[str, list[UpdateDescription]] = {}
+
+    @property
+    def filtered_updates(self) -> list[dict]:
+        """Return a list of updates that are not skipped."""
+        return [
+            {
+                "domain": domain,
+                "identifier": description.identifier,
+                "name": description.name,
+                "current_version": description.current_version,
+                "available_version": description.available_version,
+                "changelog_url": description.changelog_url,
+                "icon_url": description.icon_url,
+                "provides_backup": description.provides_backup,
+            }
+            for domain, domain_data in self._updates.items()
+            if domain_data is not None
+            for description in domain_data
+            if f"{domain}_{description.identifier}_{description.available_version}"
+            not in self._skip
+        ]
+
+    async def load(self) -> None:
+        """Load the stored data."""
+        self._skip = set(await self._store.async_load() or [])
+
+    async def gather_updates(self) -> list[dict]:
```

Comment:
Why do we store gathered updates in memory? Shouldn't we just gather those on the fly, and leave it up to the integration to decide how long to cache it ?